gc/base/SparseVirtualMemory.cpp

```diff
@@ -0,0 +1,199 @@
+#include <sys/mman.h>
+#include <sys/errno.h>
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_OSX_SPARSE_HEAP_ALLOCATION */
+
+#include "omrcomp.h"
+#include "omrport.h"
+#include "omr.h"
+
+#include "EnvironmentBase.hpp"
+#include "Forge.hpp"
+#include "GCExtensionsBase.hpp"
+#include "Math.hpp"
+#include "Heap.hpp"
+#include "HeapRegionManager.hpp"
+#include "ModronAssertions.h"
+#include "SparseVirtualMemory.hpp"
+#include "SparseAddressOrderedFixedSizeDataPool.hpp"
+
+/****************************************
+ * Initialization
+ ****************************************
+ */
+
+MM_SparseVirtualMemory*
+MM_SparseVirtualMemory::newInstance(MM_EnvironmentBase* env, uint32_t memoryCategory, MM_Heap *in_heap)
+{
+	MM_SparseVirtualMemory* vmem = NULL;
+	vmem = (MM_SparseVirtualMemory*)env->getForge()->allocate(sizeof(MM_SparseVirtualMemory), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+
+	if (vmem) {
+		new (vmem) MM_SparseVirtualMemory(env, in_heap->getPageSize(), in_heap);
+		if (!vmem->initialize(env, memoryCategory)) {
+			vmem->kill(env);
+			vmem = NULL;
+		}
+	}
+
+	return vmem;
+}
+
+bool
+MM_SparseVirtualMemory::initialize(MM_EnvironmentBase* env, uint32_t memoryCategory)
+{
+	uintptr_t in_heap_size = (uintptr_t)_heap->getHeapTop() - (uintptr_t)_heap->getHeapBase();
+	uintptr_t maxHeapSize = _heap->getMaximumMemorySize();
+	uintptr_t regionSize = _heap->getHeapRegionManager()->getRegionSize();
+	uintptr_t regionCount = in_heap_size / regionSize;
+
+	/* Ideally this should be ceilLog2, however this is the best alternative as ceilLog2 currently does not exist */
+	uintptr_t ceilLog2 = MM_Math::floorLog2(regionCount) + 1;
+
+	uintptr_t off_heap_size = (uintptr_t)((ceilLog2 * in_heap_size) / 2.0);
+#if !defined(J9ZOS390)
+	Assert_MM_true(regionSize % _pageSize == 0);
+#endif
+	bool success = MM_VirtualMemory::initialize(env, off_heap_size, NULL, NULL, 0, memoryCategory);
+
+	if (success) {
+		void *sparseHeapBase = getHeapBase();
+		_sparseDataPool = MM_SparseAddressOrderedFixedSizeDataPool::newInstance(env, sparseHeapBase, off_heap_size);
+		if (NULL == _sparseDataPool) {
+			success = false;
+		}
+	}
+
+	if (success) {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_success((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	} else {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_failure((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	}
+
+	return success;
+}
+
+bool
+MM_SparseVirtualMemory::updateSparseDataEntryAfterObjectHasMoved(void *dataPtr, void *objPtr)
+{
+	return _sparseDataPool->updateSparseDataEntryAfterObjectHasMoved(dataPtr, objPtr);
+}
+
+void *
+MM_SparseVirtualMemory::allocateSparseHeapRegionForDataAndAddDataToSparseDataPool(void *proxyObjPtr, uintptr_t size)
```

Comment:
A slighly shorter alternative to think about: allocateSparseFreeEntryAndMapToHeapObject

Reply:
You put thumb up, but did not change the name but you did change the name at another place with similar suggestion (the point of the change was mostly to use keyword 'Map' rather than 'AddData'. It's ok if like your original name more, but just making sure you did not forget, but (potentially) wanted to rename it.